src/m365/pa/commands/app/app-permission-ensure.spec.ts

```diff
@@ -0,0 +1,324 @@
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('passes validation if environmentName is used with asAdmin', async () => {
+    const actual = await command.validate({ options: { roleName: validRoleName, appName: validAppName, environmentName: validEnvironmentName, userId: validUserId, asAdmin: true } }, commandInfo);
+    assert.strictEqual(actual, true);
+  });
+
+  it('fails validation if tenant is used without role CanView', async () => {
+    const actual = await command.validate({ options: { roleName: validRoleName, appName: validAppName, tenant: true } }, commandInfo);
+    assert.notStrictEqual(actual, true);
+  });
+
+  it('passes validation if tenant is used wit role CanView', async () => {
+    const actual = await command.validate({ options: { roleName: 'CanView', appName: validAppName, tenant: true } }, commandInfo);
+    assert.strictEqual(actual, true);
+  });
+
+  it('updates permissions to a Power App by userId and sends invitation mail', async () => {
+    const postSpy = sinon.stub(request, 'post').callsFake(async (opts) => {
+      if (opts.url === `https://api.powerapps.com/providers/Microsoft.PowerApps/apps/${validAppName}/modifyPermissions?api-version=2022-11-01`) {
+        if (opts.data.put[0].properties.NotifyShareTargetOption === 'Notify'
+          && opts.data.put[0].properties.principal.id === validUserId
+          && opts.data.put[0].properties.principal.type === 'User'
+          && opts.data.put[0].properties.roleName === validRoleName) {
+          return appPermissionEnsureResponse;
+        }
+      }
+
+      throw 'Invalid request';
+    });
+
+    await command.action(logger, { options: { verbose: true, roleName: validRoleName, appName: validAppName, userId: validUserId, sendInvitationMail: true } });
+    assert(postSpy.called);
+  });
+
+  it('updates permissions to a Power App with by groupId', async () => {
+    const postSpy = sinon.stub(request, 'post').callsFake(async (opts) => {
+      if (opts.url === `https://api.powerapps.com/providers/Microsoft.PowerApps/apps/${validAppName}/modifyPermissions?api-version=2022-11-01`) {
+        if (opts.data.put[0].properties.NotifyShareTargetOption === 'DoNotNotify'
+          && opts.data.put[0].properties.principal.id === validGroupId
+          && opts.data.put[0].properties.principal.type === 'Group'
+          && opts.data.put[0].properties.roleName === validRoleName) {
+          return appPermissionEnsureResponse;
+        }
+      }
+
+      throw 'Invalid request';
+    });
+
+    await command.action(logger, { options: { verbose: true, roleName: validRoleName, appName: validAppName, groupId: validGroupId } });
+    assert(postSpy.called);
+  });
+
+  it('shares a Power App with the entire tenant', async () => {
+    auth.service.accessTokens[auth.defaultResource] = {
+      expiresOn: '123',
+      accessToken: 'abc'
+    };
```

Comment:
Can't we remove this part?

Reply:
I don't think we can.. Although we stub `getTenantIdFromAccessToken`, it still fails the test where we use the `tenant` option because `Auth.service` is undefined

Reply:
Yes correct, I was a bit too fast here.